gym_handeye/handeye.py

```diff
@@ -0,0 +1,310 @@
+    1: 'E',
+    2: 'S',
+    3: 'W',
+    4: 'G',
+    5: 'R'
+}
+
+
+class HandEye(gym.Env):
+    metadata = {'render.modes': ['human']}
+
+    def __init__(self, grid_size, note_in_hand, test_only_changes):
+        logging.debug('Starting environment HandEye')
+        self.grid_size = grid_size
+        self.note_in_hand = note_in_hand
+        self.test_only_changes = test_only_changes
+
+        self.grip_pos_x = None
+        self.grip_pos_y = None
+        self.block_in_hand = False
+        self.block_pos_x = None
+        self.block_pos_y = None
+
+        self.env_size = self.grid_size * self.grid_size + 1
+
+        self.observation_space = Discrete(self.env_size)  # camera monitors the whole grid
+        self.action_space = Discrete(6)  # N - north, S - south, W - west, E - east, G - grip, R - release
+
+        self.goal_generator_state = 0
+
+        self.observation = ['w' for x in range(self.env_size - 1)]  # array with what we observe
+        self.observation.append('0')
+
+        self.reset()
+
+    def step(self, action):
+        """
+        Run one timestep of the environment's dynamics.
+        Accepts an action and returns a tuple (observation, reward, done, info).
+        :param action: an action provided by the environment
+        :return: observation (tuple): agent's observation of the current environment
+            reward (float) : amount of reward returned after previous action
+            done (boolean): whether the episode has ended, in which case further step() calls will return undefined results
+            info (dict): contains auxiliary diagnostic information (helpful for debugging, and sometimes learning)
+        """
+
+        logging.debug('Executing a step, action = {}'.format(action))
+        previous_observation = self._observe()
+        self._take_action(action, previous_observation)
+
+        observation = self._observe()
+        reward = 0
+        episode_over = False
+        if (self.test_only_changes and previous_observation == observation) or (
+                not self.test_only_changes and previous_observation != observation):
+            episode_over = True
+
+        return tuple(observation), reward, episode_over, {}
+
+    def reset(self):
+        """
+        Resets the state of the environment and returns an initial observation.
+        Returns:
+        :return: observation (object): the initial observation of the
+            space.
+        """
+        logging.debug('Resetting the environment')
+
+        self._random_positions()
+
+        return tuple(self.observation)
+
+    def render(self, mode='human', close=False):
+        """
+        Renders the environment.
+        :param mode (str): the mode to render with
+        :param close (bool): close all open renderings
+        :return:
+        """
+        if close:
+            return
+
+        logging.debug('Rendering the environment')
+
+        if mode == 'human':
+            return self._observe()
```

Comment:
Skoro akcja środowiska odbywa się na macierzy - może ją wyrenderować a kolorem określić stan chywtaka? Będzie to bardziej naturalne do opisania stanu środowiska.